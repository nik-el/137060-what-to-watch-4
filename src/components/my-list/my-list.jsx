import React, {useEffect} from 'react';
import {Feed} from "../feed/feed";
import {Operation as FavoriteOperation} from "../../reducer/favorite";
import {useDispatch, useSelector} from "react-redux";
import {getFavorites} from "../../reducer/favorite/selectors";
import {UserBlock} from "../user-block/user-block";
import {getUser} from "../../reducer/user/selectors";
import {Logo} from "../logo/logo";

export const MyList = React.memo(function MyList({}) {
  const dispatch = useDispatch();
  const userData = useSelector(getUser);

  useEffect(() => {
    dispatch(FavoriteOperation.loadFavorites());
  }, []);

  const favoritesFilms = useSelector(getFavorites);

  return (
    <div className="user-page">
      <header className="page-header user-page__head">
        <Logo />

        <h1 className="page-title user-page__title">My list</h1>

        <UserBlock
          userData={userData}
        />
      </header>

      <section className="catalog">
        <h2 className="catalog__title visually-hidden">Catalog</h2>

        <Feed
          className="catalog__movies-list"
          currentFilms={favoritesFilms}
        />
      </section>

      <footer className="page-footer">
        <div className="logo">
          <a href="main.html" className="logo__link logo__link--light">
            <span className="logo__letter logo__letter--1">W</span>
            <span className="logo__letter logo__letter--2">T</span>
            <span className="logo__letter logo__letter--3">W</span>
          </a>
        </div>

        <div className="copyright">
          <p>Â© 2019 What to watch Ltd.</p>
        </div>
      </footer>
    </div>
  );
});
