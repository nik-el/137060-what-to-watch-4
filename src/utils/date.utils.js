const MONTHS = [
  `January`,
  `February`,
  `March`,
  `April`,
  `May`,
  `June`,
  `July`,
  `August`,
  `September`,
  `October`,
  `November`,
  `December`
];

export const getFormattedDate = (dateString) => {
  if (!dateString) {
    return ``;
  }

  const date = new Date(dateString);
  return `${MONTHS[date.getMonth()]} ${date.getDate()}, ${date.getFullYear()}`;
};
