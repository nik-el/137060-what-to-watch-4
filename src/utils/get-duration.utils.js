const HOUR_MIN = 60;
const HOUR_SEC = 3600;

export const geFormatDuration = (time) => {
  const hours = Math.floor(time / HOUR_MIN);
  const min = time % HOUR_MIN;

  return `${hours}h ${min}m`;
};

export const geVideoDuration = (time) => {
  const secNum = parseInt(time, 10);
  let hours = Math.floor(secNum / HOUR_SEC);
  let minutes = Math.floor((secNum - (hours * HOUR_SEC)) / HOUR_MIN);
  let seconds = secNum - (hours * HOUR_SEC) - (minutes * HOUR_MIN);

  if (hours < 10) {
    hours = `0` + hours;
  }
  if (minutes < 10) {
    minutes = `0` + minutes;
  }
  if (seconds < 10) {
    seconds = `0` + seconds;
  }
  return hours + `:` + minutes + `:` + seconds;
};
