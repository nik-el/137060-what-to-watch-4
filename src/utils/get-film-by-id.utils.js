export const getFilmById = (films, id) => {
  if (!Array.isArray(films) || !id) {
    return null;
  }

  return films.find((item) => item.id === id);
};
