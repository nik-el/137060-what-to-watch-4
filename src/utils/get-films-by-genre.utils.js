export const getFilmsByGenre = (films, genre) => {
  if (!Array.isArray(films)) {
    return [];
  }
  if (!genre) {
    return films;
  }

  return films.filter((item) => item.genre === genre);
};
