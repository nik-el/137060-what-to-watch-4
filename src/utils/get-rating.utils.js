export const getRating = (rating) => {
  switch (true) {
    case (rating < 3):
      return `Bad`;
    case (rating >= 3 && rating < 5):
      return `Normal`;
    case (rating >= 5 && rating < 8):
      return `Good`;
    case (rating >= 8 && rating < 10):
      return `Very good`;
    case (rating === 10):
      return `Awesome`;
    default:
      return `No rating`;
  }
};
