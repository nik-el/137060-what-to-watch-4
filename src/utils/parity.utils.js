const getOddArray = (array) => {
  if (!Array.isArray(array)) {
    return [];
  }
  return array.filter((item, index) => {
    return index % 2;
  });

};
const getEvenArray = (array) => {
  if (!Array.isArray(array)) {
    return [];
  }

  return array.filter((item, index) => {
    return !(index % 2);
  });
};

export {
  getOddArray,
  getEvenArray
};
