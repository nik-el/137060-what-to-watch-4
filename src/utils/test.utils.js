const testFilm = {
  title: `Test title 1`,
  genre: `test genre 1`,
  releaseYear: 2000,
  thumbnail: `https://picsum.photos/id/1`,
  bgImage: `https://picsum.photos/id/1`,
  bgColor: `#FFF`,
  poster: `https://picsum.photos/id/1`,
  description: `Test description`,
  rating: 5.5,
  ratingCount: 100,
  director: `Test director`,
  starring: [`Test Star 1`, `Test Star 2`],
  id: `1`,
  preview: `https://picsum.photos/id/1`,
  runTime: 100,
};
Object.freeze(testFilm);

const testEditedFilm = {
  title: `Test edit title 1`,
  genre: `test genre 1`,
  releaseYear: 2000,
  thumbnail: `https://picsum.photos/id/1`,
  bgImage: `https://picsum.photos/id/1`,
  bgColor: `#FFF`,
  poster: `https://picsum.photos/id/1`,
  description: `Test description`,
  rating: 5.5,
  ratingCount: 100,
  director: `Test director`,
  starring: [`Test Star 1`, `Test Star 2`],
  id: `1`,
  preview: `https://picsum.photos/id/1`,
  runTime: 100,
};
Object.freeze(testEditedFilm);

const testFilms = [
  {
    title: `Test title 1`,
    genre: `test genre 1`,
    releaseYear: 2000,
    thumbnail: `https://picsum.photos/id/1`,
    bgImage: `https://picsum.photos/id/1`,
    bgColor: `#FFF`,
    poster: `https://picsum.photos/id/1`,
    description: `Test description`,
    rating: 5.5,
    ratingCount: 100,
    director: `Test director`,
    starring: [`Test Star 1`, `Test Star 2`],
    id: `1`,
    preview: `https://picsum.photos/id/1`,
    runTime: 100,
  },
  {
    title: `Test title 2`,
    genre: `test genre 2`,
    releaseYear: 2000,
    thumbnail: `https://picsum.photos/id/2`,
    bgImage: `https://picsum.photos/id/2`,
    bgColor: `#FFF`,
    poster: `https://picsum.photos/id/2`,
    description: `Test description`,
    rating: 5.5,
    ratingCount: 100,
    director: `Test director`,
    starring: [`Test Star 1`, `Test Star 2`],
    id: `2`,
    preview: `https://picsum.photos/id/2`,
    runTime: 100,
  },
];
Object.freeze(testFilms);

const testGenres = [`test genre 1`, `test genre 2`];
Object.freeze(testGenres);

const testComments = [{
  id: 1,
  user: {
    id: 4,
    name: `Test  name`
  },
  rating: 1.1,
  comment: `test comment`,
  date: `2019-05-08T14:13:56.569Z`
}];
Object.freeze(testComments);

const testStore = {
  DATA: {
    films: testFilms,
    genres: testGenres,
    loadingFilmsData: false,
    restApi: `test_api_url`
  },
  VIEW: {
    currentGenre: testGenres[0],
    feedLimit: 10,
  },
  USER: {
    authorizationStatus: `TEST_AUTH`,
    user: {
      "id": 1,
      "email": `test@test.com`,
      "name": `test`,
      "avatarUrl": `test/1.png`
    }
  },
  REVIEW: {
    loadingCommentsData: false,
    loadingAddingComment: false,
    comments: testComments
  },
  FAVORITE: {
    favorites: testFilms,
    loadingFavoritesData: false,
    loadingSetFavorite: false
  }
};
Object.freeze(testStore);

export {testFilm, testFilms, testEditedFilm, testGenres, testComments, testStore};
