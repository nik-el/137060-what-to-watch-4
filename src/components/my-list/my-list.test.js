import React from 'react';
import renderer from 'react-test-renderer';

import {MyList} from './my-list';
import {Provider} from "react-redux";
import {testStore} from '../../utils/test.utils';
import configureMockStore from "redux-mock-store";
import {BrowserRouter as Router} from "react-router-dom";

const mockStore = configureMockStore();

it(`MyList renders correctly`, () => {
  const store = mockStore(testStore);

  const tree = renderer
    .create(
        <Provider store={store}>
          <Router>
            <MyList />
          </Router>
        </Provider>
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});

