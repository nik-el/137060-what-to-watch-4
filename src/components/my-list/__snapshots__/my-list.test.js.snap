// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MyList renders correctly 1`] = `
<div
  className="user-page"
>
  <header
    className="page-header user-page__head"
  >
    <div
      className="logo"
    >
      <a
        className="logo__link"
        href="/"
        onClick={[Function]}
      >
        <span
          className="logo__letter logo__letter--1"
        >
          W
        </span>
        <span
          className="logo__letter logo__letter--2"
        >
          T
        </span>
        <span
          className="logo__letter logo__letter--3"
        >
          W
        </span>
      </a>
    </div>
    <h1
      className="page-title user-page__title"
    >
      My list
    </h1>
    <a
      className="user-block"
      href="/mylist"
      onClick={[Function]}
    >
      <div
        className="user-block__avatar"
      >
        <img
          alt="User avatar"
          height="63"
          src="test_api_urltest/1.png"
          width="63"
        />
      </div>
    </a>
  </header>
  <section
    className="catalog"
  >
    <h2
      className="catalog__title visually-hidden"
    >
      Catalog
    </h2>
    <div
      className="catalog__movies-list"
    >
      <article
        className="small-movie-card catalog__movies-card"
        onClick={[Function]}
        onMouseEnter={[Function]}
        onMouseLeave={[Function]}
      >
        <div
          className="small-movie-card__image"
        >
          <img
            alt="Test title 1"
            height="175"
            src="https://picsum.photos/id/1"
            width="280"
          />
        </div>
        <h3
          className="small-movie-card__title"
          onClick={[Function]}
        >
          <a
            className="small-movie-card__link"
            href="/films/1"
            onClick={[Function]}
          >
            Test title 1
          </a>
        </h3>
      </article>
      <article
        className="small-movie-card catalog__movies-card"
        onClick={[Function]}
        onMouseEnter={[Function]}
        onMouseLeave={[Function]}
      >
        <div
          className="small-movie-card__image"
        >
          <img
            alt="Test title 2"
            height="175"
            src="https://picsum.photos/id/2"
            width="280"
          />
        </div>
        <h3
          className="small-movie-card__title"
          onClick={[Function]}
        >
          <a
            className="small-movie-card__link"
            href="/films/2"
            onClick={[Function]}
          >
            Test title 2
          </a>
        </h3>
      </article>
    </div>
  </section>
  <footer
    className="page-footer"
  >
    <div
      className="logo"
    >
      <a
        className="logo__link logo__link--light"
        href="main.html"
      >
        <span
          className="logo__letter logo__letter--1"
        >
          W
        </span>
        <span
          className="logo__letter logo__letter--2"
        >
          T
        </span>
        <span
          className="logo__letter logo__letter--3"
        >
          W
        </span>
      </a>
    </div>
    <div
      className="copyright"
    >
      <p>
        Â© 2019 What to watch Ltd.
      </p>
    </div>
  </footer>
</div>
`;
